proposals/pattern-variables.md

```diff
@@ -0,0 +1,139 @@
+	```
+   Note: This is the DeMorgan's transformation of the previous case.
+
+For a *logical_not_expression* of the form `!expr_operand`:
+- Pattern variables declared in the *expr_operand* can be redeclared anywhere in the containing expression; permitting:
+
+  ```cs
+  !(e is (0, var x)) && !(e is (var x, 0))
+  ```
+  Note: This is the expression variation of the previous case.
+
+For a *switch_case_label* of the form `case case_pattern when when_expr`:
+- Pattern variables declared in the *case_pattern* or *when_expr* can be redeclared in other case labels; permitting:
+
+	```cs
+	case (var x, 0) when e is int i:
+	case (0, var x) when a is int i:
+	```
+
+	Note: This rule is currently defined for each individual *switch_section* only.
+	
+For a *conditional_expression* of the form `expr_cond ? expr_true : expr_false`:
+- Pattern variables declared in *expr_true* can be redeclared in *expr_false*; permitting:
+
+	```cs
+	b ? x is int i : y is int i
+	```
+
+Redeclaring pattern variables is only permitted for variables of the same type.
+
+> **Open question**: How identical these types should be?
+
+Redeclaring pattern variables in any other case follows the usual scoping rules and is disallowed.
+
+### Definite assignment
+
+Note: This section is unchanged and included for the sake of completeness.
+
+For an *is_pattern_expression* of the form `e is pattern`:
+
+- The state of *v* after *is_pattern_expression* is the same as the state of *v* after *pattern*.
+
+#### General rules for pattern variables in simple patterns 
+
+Note: This section is unchanged and included for the sake of completeness.
+
+The following rules applies to any *primary_pattern* that declares a variable:
+
+- The state of *v* is definitely assigned after *primary_pattern* if the pattern is irrefutable; permitting use after:
+    
+    ```cs
+    _ = x is var x;
+    _ = 1 is int x;
+    _ = (1, 2) is var (x, y);
+    ```
+
+- Otherwise, the state of *v* is "definitely assigned when true" after *primary_pattern*; permitting use after:
+
+    ```cs
+    if (o is int x)
+    ```
+
+#### Definite assignment rules for pattern variables in pattern combinators
+
+Note: This section is derived by definite assignment rules for expressions with an exception: for patterns we strictly require both operands to assign the same set of variables. For example, even though `e is var x or 1` is logically valid, the variable `x` is not definitely assigned. That said, some of cases might still result in impossible or irrefutable patterns by other measures.
+
+For a *disjunctive_pattern* of the form `left_pattern or right_pattern`:
+
+- The definite assignment state of *v* after *disjunctive_pattern* is determined by:
+    - If the state of *v* after *left_pattern* is definitely assigned, and the state of *v* after *right_pattern* is conditionally or unconditionally definitely assigned, then the state of *v* after *disjunctive_pattern* is definitely assigned.
```

Comment:
I didn't understand why we specified "conditionally or unconditionally definitely assigned" here. That is, I don't know why `right_pattern` can conditionally assign a variable, but the variable can be fully definitely assigned after the containing `disjunctive_pattern`.

It might also be useful to specify what "conditionally definitely assigned" means. I think it just means "definitely assigned when true, or definitely assigned when false". But the spec doesn't mention the term, so it would be useful to make that completely clear.

Reply:
See https://github.com/dotnet/csharplang/pull/4592#discussion_r602993190

This portion can be removed to make it fairly identical to logical and/or expressions.

Other than that there's nothing new here. In practice, roughly speaking, we just apply binary operator rules to binary patterns.